Add top-level creation URL for global scope

TimvdLippe opened the pull request:
Global scopes have two creation URLs: one for itself and one for the "top-level" scope. It's not immediately obvious what is considered top-level here (it is not strictly defined in the specification).

In any case, reports need the creation URL of the scope itself, not the top-level version. Therefore, propagate this information from all scopes, where the worker and worklets remain to pass in `None` for their top-level scope.

Part of #37328

TimvdLippe commented on components/script/script_thread.rs:
```diff
@@ -3251,6 +3251,7 @@ impl ScriptThread {
             incomplete.viewport_details,
             origin.clone(),
             final_url.clone(),
+            Some(final_url.clone()),
```
Yes it appears to be described here: https://html.spec.whatwg.org/multipage/document-sequences.html#creating-browsing-contexts:concept-environment-top-level-creation-url We probably need to retrieve this information from `incomplete.parent_info` if it is set, otherwise set it to "about:blank"

TimvdLippe replied:
Turns out that `parent_info` doesn't seem to have the "embedder" information as I expected. `CreatorBrowsingContextInfo` appears to be related, but that doesn't have information about the surrounding element. Where would I be able to retrieve the creation URL for the embedder?

mrobinson replied:
Unfortunately, `parent_info` is probably not enough as the parent may not be the top-level frame either. You will likely need to somehow ask the constellation for this information. :/

xiaochengh reviewed (commented):
Am I understanding it correctly that the current `creation_url` actually stands for "top-level creation url" as in the spec?

If so, I suggest landing this change as 2 PRs so that it's easier to see the change clearly:
1. Rename the old `creation_url` to `top_level_creation_url`
2. Add implementation of the new `creation_url`

TimvdLippe committed:
Rename creation_url to top_level_creation_url

The URL as modelled is the top-level creation URL, since it is
None for both workers and worklets.

Signed-off-by: Tim van der Lippe <tvanderlippe@gmail.com>

TimvdLippe committed:
Add creation_url for global scopes

This is set for all global scopes, which is the URL that represents
the location of the resource. This is used in reports as a fallback
URL.

Part of #37328

Signed-off-by: Tim van der Lippe <tvanderlippe@gmail.com>

xiaochengh reviewed (commented):
The split does make it much easier to understand, thanks!

LGTH, with just a minor comment regarding the usage of options.

TimvdLippe commented on components/script/dom/window.rs:
```diff
@@ -3062,7 +3062,8 @@ impl Window {
         origin: MutableOrigin,
-        creator_url: ServoUrl,
+        creation_url: ServoUrl,
+        top_level_creation_url: Option<ServoUrl>,
```
Good catch, fixed!

TimvdLippe commented:
Seems like various security API's are unhappy about this change. Need to investigate why and maybe they should use the `creator_url` instead

TimvdLippe committed:
Align secure-context algorithm with spec

The recent fixes to the top-level creation URL unveiled
that the secure-context algorithm wasn't aligned with
the spec.

Signed-off-by: Tim van der Lippe <tvanderlippe@gmail.com>

TimvdLippe commented:
https://github.com/TimvdLippe/servo/actions/runs/15740013309

TimvdLippe commented:
https://github.com/TimvdLippe/servo/actions/runs/15742338951

TimvdLippe commented:
Yes it's all ready to go

xiaochengh commented on components/script/dom/globalscope.rs:
```diff
@@ -3431,15 +3431,30 @@ impl GlobalScope {
+            None => {
+                // Workers and worklets don't have a top-level creation URL
+                assert!(
+                    self.downcast::<WorkerGlobalScope>().is_some() ||
```
Step 1.2.1, which is for `WorkerGlobalScope`, isn't implemented here. Let's add a TODO.

> If global is a [WorkerGlobalScope](https://html.spec.whatwg.org/multipage/workers.html#workerglobalscope), then:
>

> 1. If global's [owner set](https://html.spec.whatwg.org/multipage/workers.html#concept-WorkerGlobalScope-owner-set)[0]'s [relevant settings object](https://html.spec.whatwg.org/multipage/webappapis.html#relevant-settings-object) is a [secure context](https://html.spec.whatwg.org/multipage/webappapis.html#secure-context), then return true.

> 2. Return false.

xiaochengh reviewed (approved):
LGTM with a nit.
